packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,477 @@
+   */
+  function initialize(
+    address _registry,
+    address _approver,
+    uint256 _spread,
+    uint256 _vetoPeriodSeconds
+  ) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(_registry);
+    setApprover(_approver);
+    setSpread(_spread);
+    setVetoPeriodSeconds(_vetoPeriodSeconds);
+  }
+
+  /**
+   * @notice Creates a new exchange proposal and deposits the tokens being sold.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * @param stableTokenRegistryId The string registry ID for the stable token
+   * involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The proposal identifier for the newly created exchange proposal.
+   */
+  function createExchangeProposal(
+    string calldata stableTokenRegistryId,
+    uint256 sellAmount,
+    bool sellCelo
+  ) external nonReentrant returns (uint256) {
+    address stableToken = registry.getAddressForStringOrDie(stableTokenRegistryId);
+    // Require the configurable stableToken max exchange amount to be > 0.
+    // This covers the case where a stableToken has never been explicitly permitted.
+    ExchangeLimits memory exchangeLimits = stableTokenExchangeLimits[stableTokenRegistryId];
+    require(exchangeLimits.maxExchangeAmount > 0, "Max stable token exchange amount must be > 0");
+
+    // Using the current oracle exchange rate, calculate what the buy amount is.
+    // This takes the spread into consideration.
+    uint256 buyAmount = getBuyAmount(stableToken, sellAmount, sellCelo);
+
+    // Ensure that the amount of stableToken being bought or sold is within
+    // the configurable exchange limits.
+    uint256 stableTokenExchangeAmount = sellCelo ? buyAmount : sellAmount;
+    require(
+      stableTokenExchangeAmount <= exchangeLimits.maxExchangeAmount &&
+        stableTokenExchangeAmount >= exchangeLimits.minExchangeAmount,
+      "Stable token exchange amount not within limits"
+    );
+
+    // Deposit the assets being sold.
+    IERC20 sellToken = sellCelo ? getGoldToken() : IERC20(stableToken);
+    require(
+      sellToken.transferFrom(msg.sender, address(this), sellAmount),
+      "Transfer in of sell token failed"
+    );
+
+    // Record the proposal.
+    // Add 1 to the running proposal count, and use the updated proposal count as
+    // the proposal ID. Proposal IDs intentionally start at 1 rather than 0 because
+    // LinkedList, which is used for keeping track of active proposal IDs, requires
+    // keys to be non-zero.
+    exchangeProposalCount = exchangeProposalCount.add(1);
+    exchangeProposals[exchangeProposalCount] = ExchangeProposal({
+      exchanger: msg.sender,
+      stableToken: stableToken, // for stable tokens, is saved in units to deal with demurrage.
+      sellAmount: sellCelo ? sellAmount : IStableToken(stableToken).valueToUnits(sellAmount),
+      buyAmount: buyAmount,
```

Comment:
Shouldn't this be converted to units if `sellCelo` is true?

Reply:
My rationale for a CELO -> stable token exchange is that calculating the stable token buy value at proposal-time means that the same value will be minted at execution-time, meaning the exchanger doesn't get hit by any demurrage on the stable token during that time period [proposal time, execution time].

For stable token -> CELO exchanges, an exchanger similarly isn't hit by demurrage during that time period [proposal time, execution time], because the value of the stable token at proposal time is used to calculate the buyAmount of CELO.

So the design is built to the favor of the exchanger, but imo it feels fair-- in the CELO -> stable token case, they aren't given the stable tokens until execution-time, so it doesn't seem fair to penalize them with demurrage until they're given the stable token. And in the stable token -> CELO case, they're choosing to "spend" that stable token at proposal-time, so it doesn't seem fair to penalize them after they decide to spend it.

Curious to hear your thoughts though. If you're on board I can add a comment about it too